cedar-policy-validator/src/schema_file_format.rs

```diff
@@ -203,15 +206,219 @@ pub enum SchemaType {
     },
 }
 
+impl<'de> Deserialize<'de> for SchemaType {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_any(SchemaTypeVisitor)
+    }
+}
+
+struct SchemaTypeVisitor;
+
+impl<'de> Visitor<'de> for SchemaTypeVisitor {
+    type Value = SchemaType;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("builtin type or reference to type defined in commonTypes")
+    }
+
+    fn visit_map<M>(self, mut map: M) -> std::result::Result<Self::Value, M::Error>
+    where
+        M: MapAccess<'de>,
+    {
+        #[derive(Deserialize)]
+        #[serde(field_identifier, rename_all = "camelCase")]
+        enum Fields {
+            Type,
+            Element,
+            Attributes,
+            AdditionalAttributes,
+            Name,
+        }
+
+        let mut type_name: Option<SmolStr> = None;
+        let mut element: Option<SchemaType> = None;
+        let mut attributes: Option<BTreeMap<SmolStr, TypeOfAttribute>> = None;
+        let mut additional_attributes: Option<bool> = None;
+        let mut name: Option<SmolStr> = None;
+
+        while let Some(key) = map.next_key()? {
```

Comment:
IIUC, we early deserialize the values even if their corresponding keys are deemed to be problematic after we know all keys after this loop. For instance, we still deserialize the value corresponding to `attributes` in `{"attributes": ..., "type": "Bool"}`. I wonder if we can do it lazily. For instance, store the JSON representations of the values and deserialize them if we know the key combination is correct.

Reply:
I can do this by making the variables `Option<Result<T, M::Err>>`.